### ISSUE TITLE
Cloudwatch Log Group created with options does not get auto destroyed

### ISSUE THREAD

**Issue by dwaiba** [terraform-aws-modules/terraform-aws-eks#920]
# I have issues

## I'm submitting a...

* [x] bug report
* [ ] feature request
* [x] support request - read the [FAQ](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/faq.md) first!
* [ ] kudos, thank you, warm fuzzy

## What is the current behavior?
On Creating with the following options
cluster gets created nicely but on destroy it throws 
**Error: Creating CloudWatch Log Group failed: ResourceAlreadyExistsException: The specified log group already exists:  The CloudWatch Log Group '/aws/eks/test-eks/cluster' already exists.

  on .terraform/modules/eks-cluster/terraform-aws-eks-12.1.0/cluster.tf line 1, in resource "aws_cloudwatch_log_group" "this":
   1: resource "aws_cloudwatch_log_group" "this" {**

```hcl
module "eks-cluster" {
  source                    = "terraform-aws-modules/eks/aws"
  create_eks                = lookup(var.eks_params, "createeks")
  cluster_name              = lookup(var.eks_params, "cluster_name")
  cluster_version           = lookup(var.eks_params, "cluster_version")
  subnets                   = module.vpc.public_subnets
  vpc_id                    = module.vpc.vpc_id
  cluster_enabled_log_types = ["api", "audit", "authenticator", "controllerManager", "scheduler"]

  enable_irsa = lookup(var.eks_params, "enable_irsa")

  worker_groups = [
    {
      name                 = "worker-group-1"
      instance_type        = "t2.medium"
      asg_desired_capacity = 1
      asg_max_size         = 5
      tags = [
        {
          "key"                 = "k8s.io/cluster-autoscaler/enabled"
          "propagate_at_launch" = "true"
          "value"               = "true"
        },
        {
          "key"                 = "k8s.io/cluster-autoscaler/${lookup(var.eks_params, "cluster_name")}"
          "propagate_at_launch" = "true"
          "value"               = "true"
        }
      ]
    },
    {
      name                 = "worker-group-2"
      instance_type        = "t2.medium"
      asg_desired_capacity = 2
      asg_max_size         = 5
      tags = [
        {
          "key"                 = "k8s.io/cluster-autoscaler/enabled"
          "propagate_at_launch" = "true"
          "value"               = "true"
        },
        {
          "key"                 = "k8s.io/cluster-autoscaler/${lookup(var.eks_params, "cluster_name")}"
          "propagate_at_launch" = "true"
          "value"               = "true"
        }
      ]
    }
  ]
}
```

## If this is a bug, how to reproduce? Please include a code sample if relevant.

Create with the above options and destroy

## What's the expected behavior?
The Destroy should take care of the Cloudwatch Log Group created with the options.

## Are you able to fix this problem and submit a PR? Link here if you have already.
NA
## Environment details

* Affected module version: v12.1.0
* OS:
* Terraform version:
Terraform v0.12.24
provider.aws v2.66.0
provider.kubernetes v1.11.3
provider.local v1.4.0
provider.null v2.1.2
provider.random v2.2.1
 provider.template v2.1.2

**Comment by milesarmstrong** [comment]
We're seeing this issue too. It seems that terraform does correctly destroy the log group, but that logs are still shipped to CloudWatch for some time after AWS say the EKS cluster has been deleted, causing CloudWatch to recreate the log group.

**Comment by milesarmstrong** [comment]
We're doing `terraform destroy` followed by `terraform apply` in a CI pipeline, and we've added a stage in-between the two that blocks until the log group is really gone.

We've seen it take ~3 minutes to stop being recreated.

```
function get_log_group {
  aws logs describe-log-groups --log-group-name-prefix ${LOG_GROUP_NAME} --query 'length(logGroups)'
}

REALLY_GONE=0

while [[ $REALLY_GONE -eq 0 ]]; do
  while [[ $(get_log_group) -gt 0 ]]; do
    echo "${LOG_GROUP_NAME} still exists, deleting it..."
    aws logs delete-log-group --log-group-name ${LOG_GROUP_NAME}
    sleep 5
  done
  echo "${LOG_GROUP_NAME} has been deleted. Waiting 300 seconds to validate it is not recreated..."
  sleep 300
  if [[ $(get_log_group) -eq 0 ]]; then
    REALLY_GONE=1
  fi
done
```

**Comment by maiconbaumx** [comment]
This is happening because EKS Cluster gets destroyed after Terraform delete the Cloudwatch Log Group. The AmazonEKSServicePolicy IAM policy (that is assigned to EKS Cluster role by default within this module) has permissions to CreateLogGroup and anything else needed to continue to logging correctly. When the Terraform destroys the Cloudwatch Log Group, the EKS Cluster that is running create it again. Then, when you run Terraform Apply again, the Cloudwatch Log Group doesn't exist in your state anymore (because the Terraform actually destroyed it) and the Terraform doesn't know this resource created outside him.

**Comment by stevehipwell** [comment]
Could the _Terraform_ config be changed to make sure that the _EKS_ cluster has been destroyed before the _CloudWatch_ log group is destroyed?

**Comment by barryib** [comment]
> This is happening because EKS Cluster gets destroyed after Terraform delete the Cloudwatch Log Group. The AmazonEKSServicePolicy IAM policy (that is assigned to EKS Cluster role by default within this module) has permissions to CreateLogGroup and anything else needed to continue to logging correctly. When the Terraform destroys the Cloudwatch Log Group, the EKS Cluster that is running create it again. Then, when you run Terraform Apply again, the Cloudwatch Log Group doesn't exist in your state anymore (because the Terraform actually destroyed it) and the Terraform doesn't know this resource created outside him.

Humm are you sure about that @maiconbaumx ? There is an [explicit dependency](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/cluster.tf#L44) between the cluster resource and the cloudwatch log group. Technically, Terraform should destroy `aws_cloudwatch_log_group.this` before `aws_eks_cluster.this`.

**Comment by dwaiba** [comment]
@barryib yep... that seems to be the case as in terraform-aws-modules/terraform-aws-vpc#435 too

**Comment by stevehipwell** [comment]
Is there any news on this?

**Comment by RavindraSinghKhichi** [comment]
I am facing the same issue. Any updates on this, expecting auto deletion of the log group once the VPC is destroyed.

**Comment by ulfox** [comment]
Fixed this by adding a depends_on to the cloudwatch log resource

**Comment by barryib** [comment]
@ulfox can please you share your code snippet.

**Comment by ulfox** [comment]
> @ulfox can please you share your code snippet.

Sure


    // ------------------------------------------------------------
    // EKS Cluster Cloudwatch Group
    // ------------------------------------------------------------
    resource "aws_cloudwatch_log_group" "eks_cluster" {
          count             = local.cluster_log_types_enabled ? 1 : 0
          name              = local.eks_aws_cloudwatch_log_group
          retention_in_days = local.eks_log_retention_in_days

          tags = local.eks_cloudwatch_tags
    }

    // ------------------------------------------------------------
    // EKS Cluster
    // ------------------------------------------------------------
    resource "aws_eks_cluster" "main" {
          name                      = var.eks_cluster_name
          role_arn                  = aws_iam_role.eks_cluster_role.arn
          enabled_cluster_log_types = local.cluster_log_types

          version = local.eks_version

          vpc_config {
                subnet_ids = local.eks_subnet_ids
                security_group_ids = [
                      aws_security_group.ControlPlaneSecurityGroup.id,
                ]
                endpoint_private_access = var.eks_endpoint_private_access
                endpoint_public_access  = var.eks_endpoint_public_access
          }

          timeouts {
                delete = var.eks_delete_timeouot
          }

          depends_on = [
                aws_iam_role_policy_attachment.AmazonEKSClusterPolicy,
                aws_iam_role_policy_attachment.AmazonEKSServicePolicy,
                aws_cloudwatch_log_group.eks_cluster.0
          ]
    }

**Comment by barryib** [comment]
We already have that depends_on in this module. See https://github.com/terraform-aws-modules/terraform-aws-eks/blob/v13.1.0/cluster.tf#L47

**Comment by ulfox** [comment]
> We already have that depends_on in this module. See https://github.com/terraform-aws-modules/terraform-aws-eks/blob/v13.1.0/cluster.tf#L47

apologies, I missed the module part. I thought the comment was directly for the resource.

I switched my code to use the module, cheers

**Comment by barryib** [comment]
FWIW I think that this issue is not related to terraform or this module, but it's more about how AWS buffers logs and flush them in cloudwatch log group. See also https://github.com/hashicorp/terraform/issues/14750#issuecomment-303664277.

**Comment by barryib** [comment]
I was wondering if we shouldn't deny the `logs:CreateLogGroup` for the cluster if we create the cloudwatch log group our self with Terraform.

@max-rocket-internet Any thoughts ?

**Comment by antonbabenko** [comment]
Yes, I think you are right.

If I understand correctly now the VPC flow logs will create such `logs:CreateLogGroup` if necessary (https://github.com/terraform-aws-modules/terraform-aws-vpc/blob/master/vpc-flow-logs.tf#L5). I think it should be reduced from `*` to just the necessary log group name.

There is related discussion (https://github.com/hashicorp/terraform-provider-aws/issues/902) and a note in [terraform-aws-provider/ROADMAP.md](https://github.com/hashicorp/terraform-provider-aws/blob/master/ROADMAP.md#lifecycle-retain-add-retain-attribute-to-the-terraform-lifecycle-meta-parameter).

**Comment by barryib** [comment]
But will reducing the resource from `*` to the necessary log group name avoid log group recreation by the VPC flow log service (`vpc-flow-logs.amazonaws.com`) or the EKS cluster ?

**Comment by antonbabenko** [comment]
Probably it will be working fine if permissions are scoped to the more precise resources instead of `*`, but I would recommend checking AWS docs for that.

**Comment by ostmike** [comment]
one thing that may be considered here is to allow the creation of the log group externally.

For my use case I have a cluster that I want to destroy every night and recreate every morning (development environment) however keeping the logs from previous days could be useful.

**Comment by barryib** [comment]
> one thing that may be considered here is to allow the creation of the log group externally.
> 
> For my use case I have a cluster that I want to destroy every night and recreate every morning (development environment) however keeping the logs from previous days could be useful. Perhaps using 0.13's module-level depends_on users could make sure the log group is the top of the food chain?

@ostmike Yes, we can add an option to let users create and managed their own `aws_cloudwatch_log_group`.  Please open a feature request issue for this.

I'll happy to review any PR for that.

But for sure, that would make any difference in a busy EKS environment. Because, EKS will recreate the cloudwatch log group just after Terraform destroy it, if it has logs left.

**Comment by den-is** [comment]
That's still an issue.
You have to "manually" destroy CloudWatch Log group if you want to destroy existing cluster and resources and recreate cluster with exactly same name, rather than creating cluster with some random new name.

**Comment by navaati** [comment]
Wowowow, hol’up Mr. Bot, don’t stale, this is still an issue ! So yes as @barryib said the solution, same as for the VPC module which had the same problem, is to deny the logs:CreateLogGroup permission in the service role: it is not needed since we create the log group ourselves with Terraform, and it is causing trouble. See the implementation for the VPC module: https://github.com/terraform-aws-modules/terraform-aws-vpc/pull/550. @antonbabenko you looked at it the other day so you may remember the details.

Now the situation here is a tad more complex because we don’t create the policy ourselves, we use the AWS provided one `AmazonEKSClusterPolicy` so it’s not like we can apply the exact same patch as for the VPC module. The solution might take the form of adding a specific deny on top of the AWS policy (which we still want to use to benefit from it always being up to date with EKS). Not so sure yet how to do that, @barryib is this what you were thinking about, do you know how to do it ?

**Comment by barryib** [comment]
@navaati that was the magic. Can you please open a PR for that ? Could you please test it also ?

**Comment by navaati** [comment]
Hi.

As I described in my previous message I have a vague idea of what’s needed but I don’t know enough to throw a patch in the air and hope for someone else to test it, and I switched jobs so I’m not really in a position to develop/test this anymore, sorry :/.

I’ll keep an eye on the issue and I might decide to invest some personal time and bucks regarding this bug but that’s very hypothetical at this point, because I don’t use EKS at home or on my new job.

**Comment by navaati** [comment]
Not stale.

**Comment by haarchri** [comment]
we tested it with #1594 and deny the create log group - we managed to get the auto destroy working please have a look 
thanks guys

**Comment by antonbabenko** [comment]
This issue has been resolved in version 17.24.0 :tada:

**Comment by mikeroll** [comment]
Having the same symptoms when destroying and recreating the cluster on `v18.4.1`. I can't seem to find the IAM hack introduced by #1594 anymore - was it lost during the v18 update or removed intentionally for some reason?

**Comment by bryantbiggs** [comment]
> Having the same symptoms when destroying and recreating the cluster on `v18.4.1`. I can't seem to find the IAM hack introduced by #1594 anymore - was it lost during the v18 update or removed intentionally for some reason?

This was tested and deemed no longer required and therefore removed in v18 - I test this module quite often throughout the week for various PRs and do not see any issues with cloudwatch logs not deleting. please feel free to open a new issue with reproduction steps if necessary

**Comment by daroga0002** [comment]
> This was tested and deemed no longer required and therefore removed in v18 - I test this module quite often throughout the week for various PRs and do not see any issues with cloudwatch logs not deleting. please feel free to open a new issue with reproduction steps if necessary

@bryantbiggs you cannot replicate this because you dont have a workload over cluster. This is caused by general latency of cloudwatch log processing under huge load. EKS plane producing logs and is destroyed, cloudwatch group is also destroyed and then cloudwatch internal AWS queue recreates log group (when it was arlelady removed by terraform)

**Comment by bryantbiggs** [comment]
> > This was tested and deemed no longer required and therefore removed in v18 - I test this module quite often throughout the week for various PRs and do not see any issues with cloudwatch logs not deleting. please feel free to open a new issue with reproduction steps if necessary
> 
> 
> 
> @bryantbiggs you cannot replicate this because you dont have a workload over cluster. This is caused by general latency of cloudwatch log processing under huge load. EKS plane producing logs and is destroyed, cloudwatch group is also destroyed and then cloudwatch internal AWS queue recreates log group (when it was arlelady removed by terraform)
> 
> 
> 
> 
> 
> 

If that's the case, it's still not the concern of the module. It sounds like users need to delete/stop workloads before destroying the cluster

**Comment by haarchri** [comment]
@[bryantbiggs](https://github.com/bryantbiggs) why then the overlay deny was removed with version V18 ? when the only option is now to say we need delete/stop workloads before - because we delivered the fix that it fits

**Comment by bryantbiggs** [comment]
I believe I answered this above. Users can attach additional policies as needed so feel free to do so with your configuration if that is what you need

**Comment by timblaktu** [comment]
I hit this recently on a simple eks cluster with virtually no workload running when I `terraform destroy`ed it. I was reading through the issues on this, none of which really seem to completely fix it so I'm just going to manually delete the already-existing resources for now and move on. FWIW, in my case, I'm reproduced this while creating/destroying the cluster in fairly rapid succession.

**Comment by trallnag** [comment]
@timblaktu, why is the explicit deny "hack" not working for you?

**Comment by timblaktu** [comment]
I don't know. We're using v18.x so we should have the deny hack (you mean [this](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1594), right?) in place, since it was introduced in 17.24.0. In my case, this is not easily reproducible, and I'm imagining it happens in extraordinary circumstances, like maybe iff a terraform process (apply or destroy) aborted ungracefully or something..

**Comment by timblaktu** [comment]
@bryantbiggs I upgraded our module to 18.17.1 and will start a tight loop creating/destroying our infra. We don't run much workloads either yet and that might be why this is so rare. Ran into it again today, and suspect it's in cases where we destroy immediately after create finishes..

**Comment by sirex99** [comment]
yes, this is still an issue with 18.29.0

**Comment by sirex99** [comment]
@bryantbiggs just to confirm i'm also seeing this issue on the latest version. Could we get the code of https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1594 re-added again ?

**Comment by bryantbiggs** [comment]
https://github.com/terraform-aws-modules/terraform-aws-eks/blob/7f901840c430c0b631cc4066a87779b915a07524/main.tf#L266-L287

**Comment by bryantbiggs** [comment]
perhaps the ARN for the CloudWatch log group is not sufficient for this type of denial policy. This has been changed back to what the original PR had with a wildcard `"*"` and is available in v18.30.1 https://github.com/terraform-aws-modules/terraform-aws-eks/pull/2267
